src/test/app/Wasm_test.cpp

```diff
@@ -695,6 +695,1089 @@ struct Wasm_test : public beast::unit_test::suite
+                &hfs,
+                1000000);
+            (void)re;  // Testing crash resistance
+            // Don't enforce success/failure - just ensure it doesn't crash
+        }
+
+        // Test CPU exhaustion through loops
+        {
+            // Create a simple infinite loop WASM module
+            Bytes infiniteLoopWasm = {
+                0x00, 0x61, 0x73, 0x6D, 0x01, 0x00, 0x00, 0x00,  // WASM header
+                0x01, 0x04, 0x01, 0x60, 0x00, 0x00,  // Type section: () -> ()
+                0x03, 0x02, 0x01, 0x00,              // Function section
+                0x07, 0x08, 0x01, 0x04, 0x6C, 0x6F, 0x6F, 0x70,
+                0x00, 0x00,  // Export "loop"
+                0x0A, 0x06, 0x01, 0x04, 0x00, 0x03, 0x40, 0x0C,
+                0x00, 0x0B  // Code: loop { br 0 }
+            };
+
+            auto& engine = WasmEngine::instance();
+            {
+                auto re =
+                    engine.run(infiniteLoopWasm, "loop", {}, {}, nullptr, 10);
+                BEAST_EXPECT(!re.has_value());
+            }
+            {
+                auto re = engine.run(
+                    infiniteLoopWasm, "loop", {}, {}, nullptr, 1'000'000);
+                BEAST_EXPECT(!re.has_value());
+            }
+        }
+
+        // Test memory boundary attacks
+        {
+            auto const wasmStr =
+                boost::algorithm::unhex(allHostFunctionsWasmHex);
+            Bytes const wasm(wasmStr.begin(), wasmStr.end());
+
+            // Test with host function that returns data at exact memory
+            // boundaries
+            struct BoundaryTestHostFunctions : public TestHostFunctions
+            {
+                mutable int callCount = 0;
+                explicit BoundaryTestHostFunctions(Env& env)
+                    : TestHostFunctions(env)
+                {
+                }
+
+                Expected<Bytes, HostFunctionError>
+                getTxField(SField const& fname) override
+                {
+                    callCount++;
+                    switch (callCount)
+                    {
+                        case 1:
+                            return Bytes(0, 0);  // Empty
+                        case 2:
+                            return Bytes(1, 0xFF);  // Minimal
+                        case 3:
+                            return Bytes(4096, 0xAA);  // Page boundary
+                        default:
+                            return TestHostFunctions::getTxField(fname);
+                    }
+                }
+            };
+
+            BoundaryTestHostFunctions boundaryHfs(env);
+            auto re = runEscrowWasm(
+                wasm, ESCROW_FUNCTION_NAME, {}, &boundaryHfs, 100000);
+            BEAST_EXPECT(
+                re.has_value() ||
+                !re.has_value());  // Either result is acceptable
+            // Should handle boundary conditions gracefully
+        }
+
+        // Test concurrent-like behavior simulation
+        {
```

Comment:
This test doesn't test concurrent behavior. Run it with parallel threads.  Multiple sequential runs already achieved by all the other tests ( that's how they call wasm engine)